pallets/pallet-bonded-coins/src/benchmarking.rs

```diff
@@ -1 +1,855 @@
 
+use crate::{
+	Call, CollateralAssetIdOf, CollateralCurrenciesBalanceOf, Config, CurveParameterTypeOf, FungiblesAssetIdOf, Pallet,
+};
+
+pub trait BenchmarkHelper<T: Config> {
+	fn calculate_collateral_asset_id(seed: u32) -> CollateralAssetIdOf<T>;
+
+	fn calculate_bonded_asset_id(seed: u32) -> FungiblesAssetIdOf<T>;
+}
+
+impl<T> BenchmarkHelper<T> for ()
+where
+	T: Config,
+	CollateralAssetIdOf<T>: From<u32>,
+	FungiblesAssetIdOf<T>: From<u32>,
+{
+	fn calculate_collateral_asset_id(seed: u32) -> CollateralAssetIdOf<T> {
+		seed.into()
+	}
+
+	fn calculate_bonded_asset_id(seed: u32) -> FungiblesAssetIdOf<T> {
+		seed.into()
+	}
+}
+
+#[benchmarks(where
+	<CurveParameterTypeOf<T> as Fixed>::Bits: Copy + ToFixed + AddAssign + BitOrAssign + ShlAssign + TryFrom<U256>,
+	CollateralCurrenciesBalanceOf<T>: Into<U256> + TryFrom<U256>,
+	<T as Config>::CollateralCurrencies: Create<T::AccountId> ,
+	<T as Config>::Fungibles: InspectRoles<T::AccountId> + AccountTouch<FungiblesAssetIdOf<T>, AccountIdOf<T>>,
+	<T as Config>::DepositCurrency: Mutate<T::AccountId>,
+	<T as Config>::CollateralCurrencies: MutateFungibles<T::AccountId>,
+	AccountIdLookupOf<T>: From<T::AccountId>,
+)]
+mod benchmarks {
+	use frame_support::traits::OriginTrait;
+	use frame_support::traits::{
+		fungible::{Inspect, Mutate, MutateHold},
+		fungibles::{Destroy, Inspect as InspectFungibles, Mutate as MutateFungibles},
+	};
+	use frame_support::traits::{fungibles::Create, AccountTouch, EnsureOrigin, Get};
+	use sp_runtime::{traits::Zero, BoundedVec, SaturatedConversion};
+	use sp_std::ops::Mul;
+
+	use crate::{
+		curves::Curve,
+		mock::*,
+		types::{Locks, PoolManagingTeam, PoolStatus},
+		AccountIdLookupOf, AccountIdOf, CollateralAssetIdOf, CurveParameterInputOf, HoldReason, PoolDetailsOf, Pools,
+		TokenMetaOf,
+	};
+
+	use super::*;
+
+	// helper functions
+	// collateral currencies
+	fn create_collateral_asset<T: Config>(asset_id: CollateralAssetIdOf<T>)
+	where
+		<T as Config>::CollateralCurrencies: Create<T::AccountId>,
+	{
+		let pool_account = account("collateral_owner", 0, 0);
+		<T as Config>::CollateralCurrencies::create(asset_id, pool_account, false, 1u128.saturated_into())
+			.expect("Creating collateral asset should work");
+	}
+
+	fn calculate_default_collateral_asset_id<T: Config>() -> CollateralAssetIdOf<T> {
+		T::BenchmarkHelper::calculate_collateral_asset_id(u32::MAX)
+	}
+
+	fn create_default_collateral_asset<T: Config>() -> CollateralAssetIdOf<T>
+	where
+		<T as Config>::CollateralCurrencies: Create<T::AccountId>,
+	{
+		let collateral_id = calculate_default_collateral_asset_id::<T>();
+		create_collateral_asset::<T>(collateral_id.clone());
+		collateral_id
+	}
+
+	fn set_collateral_balance<T: Config>(asset_id: CollateralAssetIdOf<T>, who: &AccountIdOf<T>, amount: u128)
+	where
+		<T as Config>::CollateralCurrencies: MutateFungibles<T::AccountId>,
+	{
+		T::CollateralCurrencies::set_balance(asset_id, who, amount.saturated_into());
```

Comment:
Should you also `assert` after this, to make sure it worked?

Reply:
I added some asserts in: [358f418](https://github.com/KILTprotocol/kilt-node/pull/789/commits/358f418d0dafb9d36a930edfeec5019aa970c6ae)